notes/notes.go

```diff
@@ -0,0 +1,345 @@
+		HideFromHelp:        true,
+		RequireDiscordPerms: requiredPerms,
+		CommandType:         discordgo.UserApplicationCommand,
+		IsResponseEphemeral: true,
+		RunFunc: func(data *dcmd.Data) (interface{}, error) {
+			notes, err := getNotes(data.Context(), data.GuildData.GS.ID, data.SlashCommandTriggerData.TargetID)
+			if err != nil {
+				return nil, err
+			}
+			return notes.createMessage(), nil
+		},
+	},
+}
+
+func getNotes(ctx context.Context, guildID, userID int64) (*parsedNotes, error) {
+	entry, err := models.Notes(qm.Where("guild_id = ? AND member_id = ?", guildID, userID)).OneG(ctx)
+	if err != nil && err != sql.ErrNoRows {
+		return nil, err
+	}
+
+	var notes []*Note
+	if entry != nil {
+		for i := 0; i < maxNotesPerUser && i < len(entry.Notes); i++ {
+			notes = append(notes, parseRawNote(guildID, entry.Notes[i]))
+		}
+	}
+
+	userNotes := &parsedNotes{
+		guildID: guildID,
+		userID:  userID,
+		notes:   notes,
+	}
+
+	return userNotes, nil
+}
+
+func (p *parsedNotes) add(note string, author *discordgo.User) error {
+	var index *int
+	for i, n := range p.notes {
+		if len(n.noteLines) < 1 || n.noteLines[0] == "" {
+			index = &i
+			break
+		}
+	}
+
+	if index == nil {
+		i := len(p.notes)
+		if i >= maxNotesPerUser {
+			return commands.NewUserError("This user has the max amount of notes! Please delete or edit one.")
+		}
+		p.notes = append(p.notes, &Note{})
+		index = &i
+	}
+
+	return p.edit(*index, note, author)
+}
+
+func (p *parsedNotes) edit(index int, note string, author *discordgo.User) error {
+	if len(p.notes) <= index {
+		// this shouldn't be possible
+		logger.WithField("guild", p.guildID).WithField("index", index).WithField("length", len(p.notes)).Warn("a custom ID tried to select a nonexistent note")
+		return invalidSelectionErr
+	}
+	note = strings.TrimSpace(note)
+	lines := strings.Split(note, "\n")
+	now := time.Now()
+	p.notes[index] = &Note{
+		noteLines: lines,
+		signature: author.String(),
+		authorID:  author.ID,
+		updatedAt: now,
+	}
+	return nil
+}
+
+func (p *parsedNotes) delete(index int) error {
+	if len(p.notes) <= index {
+		// this shouldn't be possible
+		logger.WithField("guild", p.guildID).WithField("index", index).WithField("length", len(p.notes)).Warn("a custom ID tried to select a nonexistent note")
+		return invalidSelectionErr
+	}
+	p.notes[index] = &Note{}
+	return nil
+}
+
+func (p *parsedNotes) save(ctx context.Context) error {
+	var formattedNotes []string
+	for _, n := range p.notes {
+		formattedNotes = append(formattedNotes, n.formatNote())
+	}
+
+	logger.Info(formattedNotes)
+
+	model := models.Note{
+		GuildID:  p.guildID,
+		MemberID: p.userID,
+		Notes:    formattedNotes,
+	}
+
+	whitelist := boil.Whitelist("notes")
+	return model.Upsert(ctx, common.PQ, true, []string{"guild_id", "member_id"}, whitelist, boil.Infer())
+}
+
+const noNotesMsg = "No shared staff notes written for this user."
+
+func (p *parsedNotes) createMessageContent() *discordgo.MessageSend {
```

Comment:
This is a minor nit, but I'm concerned `parsedNotes` is juggling too many concerns at the moment. It's responsible for holding parsed notes (good), for saving and modifying the corresponding database entry (good), but seemingly also for Discord-related things like generating message content, custom IDs, buttons, messages, and modals. I think those last few items can be separated into free-standing functions that accept `parsedNotes` as an argument.

Concretely, I'm wondering if

```go
type Notes struct {
    guildID, userID int
    notes []Note
}

func (*Notes) Add(...)
func (*Notes) Edit(...)
func (*Notes) Save(...)

func CreateMessageContent(*Notes)
func CreateButtons(*Notes)
// ...
```
would be a better organizational structure for this code.

Reply:
Yep that makes sense to me, will adjust a little later.